exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,305 @@
+#include <tfc/confman.hpp>
+#include <tfc/confman/observable.hpp>
+#include <tfc/dbus/string_maker.hpp>
+
+namespace asio = boost::asio;
+namespace details = tfc::ipc::details;
+
+// File under /etc/tfc/mqtt-broadcaster/def/mqtt_broadcaster.json which lists the signals that are not supposed to be
+// published on the mqtt broker. Updating the signals in this file will cause the running program to stop and restart
+struct config {
+  tfc::confman::observable<std::vector<std::string>> _banned_topics{};
+  struct glaze {
+    static constexpr auto value{ glz::object("_banned_topics", &config::_banned_topics) };
+    static constexpr auto name{ "mqtt_broadcaster" };
+  };
+};
+
+// In order to ease testing, types are injected into the class
+template <class ipc_client_type, class mqtt_client_type>
+class mqtt_broadcaster {
+public:
+  mqtt_broadcaster(asio::io_context& ctx,
+                   std::string mqtt_address,
+                   std::string mqtt_port,
+                   std::string mqtt_username,
+                   std::string mqtt_password,
+                   ipc_client_type& ipc_client,
+                   std::shared_ptr<mqtt_client_type> mqtt_client)
+      : object_path_(tfc::dbus::make_dbus_path("ipc_ruler")), interface_name_(tfc::dbus::make_dbus_name("manager")),
+        mqtt_host_(std::move(mqtt_address)), mqtt_port_(std::move(mqtt_port)), mqtt_username_(std::move(mqtt_username)),
+        mqtt_password_(std::move(mqtt_password)), ctx_(ctx), ipc_client_(ipc_client), mqtt_client_(std::move(mqtt_client)),
+        logger_("mqtt_broadcaster"),
+        dbus_connection_(std::make_unique<sdbusplus::asio::connection>(ctx, tfc::dbus::sd_bus_open_system())),
+        signal_updates_(std::make_unique<sdbusplus::bus::match::match>(
+            *dbus_connection_,
+            sdbusplus::bus::match::rules::propertiesChanged(object_path_, interface_name_),
+            std::bind_front(&mqtt_broadcaster::update_signals, this))),
+        config_(ctx, "mqtt_broadcaster") {
+    asio::co_spawn(mqtt_client_->strand(), initialize(), asio::detached);
+  }
+
+private:
+  auto initialize() -> asio::awaitable<void> {
+    // Initial connect to the broker
+    if (!connect_active_) {
+      connect_active_ = true;
+      co_await asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::use_awaitable);
+    }
+
+    // When a new list of banned string arrives the program is restarted
+    config_.value()._banned_topics.observe([this](auto& new_conf, [[maybe_unused]] auto& old_conf) {
+      banned_signals_.clear();
+      for (auto const& con : new_conf) {
+        banned_signals_.push_back(con);
+      }
+      restart();
+    });
+
+    asio::co_spawn(mqtt_client_->strand(), tfc::base::exit_signals(ctx_), asio::detached);
+
+    banned_signals_ = config_.value()._banned_topics.value();
+    load_signals();
+  }
+
+  // Connect the mqtt client to the broker. Only a single "thread" can run this at a time. The routine that calls this
```

Comment:
it is not really a multi thread problem, this routine can only be spawned once.

Reply:
you probably could change the connect flag to a strand to make this safe

Reply:
> it is not really a multi thread problem, this routine can only be spawned once.

How about changing the comment to:

```cpp
  // Connect the mqtt client to the broker. This function can only be spawned once to avoid multiple spawns to open/close the
  // socket at the same time. The routine that calls this function must set the connect_active_ flag to true before calling
  // this function and the function sets it to false when finished.
```

Reply:
> you probably could change the connect flag to a strand to make this safe

not sure how to do this but will look into it. would appreciate suggestions.

Reply:
> > you probably could change the connect flag to a strand to make this safe
> 
> not sure how to do this but will look into it. would appreciate suggestions.

if the class would own a connecting_strand member you could post the connect method on the strand and in the beginning of the function check if you are already connected

Reply:
> > > it is not really a multi thread problem, this routine can only be spawned once.
> > 
> > 
> > How about changing the comment to:
> > ```c++
> >   // Connect the mqtt client to the broker. This function can only be spawned once to avoid multiple spawns to open/close the
> >   // socket at the same time. The routine that calls this function must set the connect_active_ flag to true before calling
> >   // this function and the function sets it to false when finished.
> > ```
> 
> seems okay, but why is the flag not set in the function itself, moving the responsibility of the flag to the function

done